plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/storage/TrieBasedStore.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.wlm.rule.storage;
+
+import org.apache.commons.collections4.trie.PatriciaTrie;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This is a patricia trie based implementation of AttributeValueStore
+ */
+public class TrieBasedStore implements AttributeValueStore {
+    PatriciaTrie<String> trie;
+
+    public TrieBasedStore(PatriciaTrie<String> trie) {
+        this.trie = trie;
+    }
+
+    @Override
+    public void addValue(String key, String value) {
+        trie.put(key, value);
+    }
+
+    @Override
+    public void removeValue(String key) {
+        trie.remove(key);
+    }
+
+    @Override
+    public Optional<String> getValue(String key) {
+        /**
+         * Since we are inserting prefixes into the trie and searching for larger strings
+         * It is important to find the largest matching prefix key in the trie efficiently
+         * Hence we can do binary search
+         */
+        String longestMatchingPrefix = findLongestMatchingPrefix(key);
```

Comment:
Should we try doing a `get()` first directly on the key in case we get a exact match and return? Assuming it might be faster(in some cases) instead of doing binary search and calling `prefixMap()` to extract longestMatchingPrefix.

Reply:
`prefixMap` will also have the same complexity for larger keys since the search is guided. IMO it will only clutter the method